core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1808,7 +1809,10 @@ class ReplicaManager(val config: KafkaConfig,
 
           // pause cleaning for partitions that are being moved and start ReplicaAlterDirThread to move
           // replica from source dir to destination dir
-          logManager.abortAndPauseCleaning(topicPartition)
+          replicaAlterLogDirsManager.getFetcher(topicPartition) match {
```

Comment:
Can you please explain the reasoning behind introducing this? I understand that you are trying to save a call to abortAndPauseCleaning, but I do not understand why

Reply:
In general, we should pause cleaning only once during altering log dir. Because there is a counter (kafka.log.LogCleaningPaused) for how many times the cleaning of topic partition has been paused and the fetcher only subtracts the counter by one after altering completed. And the cleaning will only be resumed when the counter is reduced to 0. But currently every time broker recives a LeaderAndISR request about the parition being altering log dir, it would call the abortAndPauseCleaning. Therefore, during altering log dir, a LeaderAndISR about that partition would disable the log cleaning of that partition forever.

Reply:
Recently, I deployed this change to my cluster and found that this change didn't work in the scenario where an exception was raised. During altering, the partition being altered log dir would be marked as failed due to exceptions and the fetcher of it would be removed **without resuming the cleaning**. After receiving a LeaderAndISR request of that partition, the altering would be restarted by assign a fetcher to that partition, and the cleaning of that partition would be paused one more time due to no assigned fetcher. Therefore, the cleaning will never be resumed (pause twice, but resume once after altering completed).

I pushed another commit. Every time assign fetcher to partition, pause the cleaning of that partition. And every time remove the fetcher, resume the cleaning of that partition. To make sure that resuming as many times as pausing. If this change make sense to you, I will add unit test later.

Reply:
Hi @clolov, could you take a look?

Reply:
Hi @divijvaidya, could you take a look?

Reply:
Sorry, I won't have time anytime soon to look into this.

Reply:
@clolov @divijvaidya Kindly ping.